katsdptelstate/redis.py

```diff
@@ -167,13 +212,17 @@ def monitor_keys(self, keys):
                         timeout = yield (key,)
                     elif message['type'] == 'message':
                         # TODO: eventually change the protocol to indicate
-                        # mutability in the message.
+                        # the key type in the message.
                         type_ = self.client.type(key)
                         if type_ == b'string':
                             timeout = yield (key, message['data'])
-                        else:
+                        elif type_ == b'hash':
+                            timeout = yield (key,)
```

Comment:
It doesn't seem massively unreasonable to include the name of the first sub-key changed in the message. I can imagine progressively updated hashes (say single shot quality metrics that get generated over time), and someone interested in waiting for a particular sub-key to get set with the metric they care about.

Reply:
The current design of `wait_key` is that the predicate is given the result of `get`, and for indexed keys `get` returns the whole dict. Do you have an idea for how `wait_key` should work for indexed keys to support efficiently waiting on a subkey?

It's definitely something we should get right the first time, because updating the pubsub protocol is a backwards compatibility nightmare.

Reply:
From the client side I would say allowing a sub_key to be specified in the wait_key call that passes key[sub_key] to the condition when the sub_key is updated.

Since we use HSETNX in the call we are only ever updating one sub_key at a time, so just including this sub_key in the message and then parsing this out in monitor_keys seems relatively straightforward.

If we allow bulk updating at some future date, then we would need to send all sub_key changes in the message, so we should probably cater for this from the start and send a list in the update message.

Reply:
> From the client side I would say allowing a sub_key to be specified in the wait_key call that passes key[sub_key] to the condition when the sub_key is updated.

Makes sense. What should it do if the key isn't indexed?

I'll have to think about implementation, because I'll need to bundle the subkey and value into a single pubsub message. Redis' Lua implementation comes with a msgpack library, but I don't think that'll work for fakeredis so I'll need to either extend fakeredis or some up with a simpler encapsulation.

Reply:
If you have asked to match a key/sub-key pair and the key happens to be some other type then the wait should just timeout without matching. 

Do you need to include the value in the pubsub message - I know it is a little less efficient but why not just look it up on the client side when you get the message about the changed key. This also stops big values accidentally being sent out on the pubsub channels.

Reply:
I've implemented a `wait_indexed` method - it was sufficiently different to `wait_key` that I didn't want to try to shoe-horn the functionality in there, and it's analogous to `get_indexed`. Since calling it is an implicit assertion that the key is indexed, I've made it an ImmutableKeyError if it isn't.

I don't see that it makes a big performance difference whether a big value is sent to the client over the pub-sub connection or in the subsequent request it makes for the value. For mutable values it avoids a race condition where the value might have changed again before the client got around to asking for it. It's also very difficult to change the pubsub protocol (because you have to change all clients simultaneously).